### ISSUE TITLE
Publish fails when using HTTPClient

### ISSUE THREAD

**Issue by BernH4** [knolleary/pubsubclient#892]
By reducing my code to a minimum i found out that client.publish fails without any error when the following code is called **before**:
```c
  HTTPClient httpClient;
  httpClient.begin(espClient, "google.de" );
  httpClient.end();
```
As im still learning please let me know if i could somehow debug this on my own any further.
Heres a minimum working example, wifi and mqtt config has to be updated:

```
#include <ESP8266WiFi.h>
#include <ESP8266HTTPClient.h>
#include <PubSubClient.h>
#include "credentials.h" //My Wifi and MQTT credentials

const char* ssid = WIFI_SSID;  //CHANGE TO YOUR OWN CONFIG
const char* password = WIFI_PASSWD;//CHANGE TO YOUR OWN CONFIG
const char* mqttServer = "192.168.178.143";//CHANGE TO YOUR OWN CONFIG
const int mqttPort = 1883;//CHANGE TO YOUR OWN CONFIG
const char* mqttUser = MQTT_USERN;//CHANGE TO YOUR OWN CONFIG
const char* mqttPassword = MQTT_PASSWD;//CHANGE TO YOUR OWN CONFIG

WiFiClient espClient;
PubSubClient client(espClient);

void setup() {
  Serial.begin(115200);

  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, password);

  while (WiFi.status() != WL_CONNECTED) {
    delay(5);
    Serial.print(WiFi.status());
  }

  client.setServer(mqttServer, mqttPort);
  while (!client.connected()) {
    Serial.println("Connecting to MQTT...");
    if (client.connect("ESP8266Client", mqttUser, mqttPassword )) {
      Serial.println("connected");
    }
    else {
      Serial.print(client.state());
      delay(5);
    }
  }

  HTTPClient httpClient;
  httpClient.begin(espClient, "google.de" );
  httpClient.end();

  Serial.println("now publishing...");

  client.publish("debug/debug", "hello");

  Serial.println("should now be published!");
  delay(100);
}

void loop(){}

```

**Comment by jpdw** [comment]
No solution (sorry), but I think I am seeing the same as you.

When I initially built your code, the publish worked every time. Thiswas in my rarely-used Arduino install, and I had had to slightly modify the httpClient.begin call without the espClient param -- indicating that I've got an old version of the libraries there.  Building it again in PlatformIO (which I use normally and is up-to-date) I get the same failure as you.  Not sure how that helps but might be something to investigate.

These two cases seem to be similar: [stack exchange](https://arduino.stackexchange.com/questions/86374/esp8266-how-to-run-mqtt-client-and-http-client-in-one-loop) and [esp8266 forum](https://www.esp8266.com/viewtopic.php?f=12&t=21353).

Given the stability of PubSubClient my current guess is a change in the esp8266 libs has broken something that used to work.

**Comment by jpdw** [comment]
OK -- I think I've found the solution - at least, it seem now to work for me.

TL;DR - you need separate WiFiClient objects for each connection.

The naming "WiFiClient" is perhaps a bit mis-leading -- we only have 1 x WiFi interface on the esp (or most other devices). The class name was probably to match a pre-existing Arduino class. Reading the [Arduino docs](https://www.arduino.cc/en/Reference/WiFiClient) was the clue:

> Creates a client that can **connect to to a specified internet IP address and port** as defined in client.connect().

It's a single connection to a single socket (port). So if it's been initialised to 1883 and pub/sub-ing for mqtt, the moment http.GET() is called, it is re-used for port 80 and consequently mqtt is broken. The fact that you can do http.begin() without mqtt stopping adds some confusion, but presumably nothing is actually "begun" until http.GET()/PUT() etc are called.

"Obvious really!!" Once you know, that is!

In your example, change:
```
WiFiClient espClient;
```

to

```
WiFiClient espClient_mqtt;
WiFiClient espClient_http;
```
and then update the espClient name where you initialise `client` and call `httpClient.begin` and it should now work.

**Comment by BernH4** [comment]
@jpdw Thanks so much for your time, it's working perfectly now :)

**Comment by tjkolev** [comment]
@jpdw,
Late to the party. In my case all http calls after the first one were failing with HTTPC_ERROR_CONNECTION_LOST (from HTTPClient). Spent close to a day troubleshooting...
Thank you!
